JPEG XL experimental support using j40

qbnu opened the pull request:
Adds partial support for JPEG XL using the [J40 decoder](https://github.com/lifthrasiir/j40). Does not support animations, losslessly transcoded JPEGs, some compression methods, or advanced features. Images encoded by [cjxl](https://github.com/libjxl/libjxl/releases/tag/v0.7.0) with `-q 100` (lossless) and `-e 9` (max effort) will either fail to load or [display incorrectly](https://user-images.githubusercontent.com/93988953/206628486-308251bb-0eaf-4290-8483-4ee791ce210a.png) ([original](https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png)).

Compared to [ImageMagick Display](https://imagemagick.org/script/download.php#windows), which [uses libjxl with multithreading](https://github.com/ImageMagick/ImageMagick/blob/main/coders/jxl.c), it loads lossless JXL images 30-100% slower, and lossy images ~160% slower.

sylikc commented:
did it not work?

qbnu commented:
I managed to get libjxl to work instead, which is MUCH faster than I thought compared to j40. The downside is that it requires more prebuilt binaries and the Visual C++ Redistributable. The branch is [here](https://github.com/qbnu/jpegview/tree/jpeg-xl-libjxl). 
[JPEGView-libjxl.zip](https://github.com/sylikc/jpegview/files/10241807/JPEGView-libjxl.zip)

sylikc commented:
that might be alright.  I can submodule it into the `extras` directory so there's tracking as to which source the libjxl comes from.  Then use the JXLWrapper that you wrote.  Does it need all the DLL files?  Could certainly build the DLLs in the submodules directory then copy them over to the distribution.

It could work like WEBP where, even if the WEBP.dll doesn't exist, JPEGView will run, but won't have WEBP support.

qbnu commented:
It doesn't need `brotlienc.dll` so I removed that, and I added `jxl_dec.dll` and `jxl_threads.dll` to delay loaded DLLs so you don't need them to use JPEGView. What do you think should be done about the Visual C++ Runtime?
- Include `vcruntime140.dll` and `msvcp140.dll` in the folder
- Bundle it with the installer
- Tell the user to go download it if they want JPEG XL support

sylikc commented:
So, Microsoft allows VC++ Runtime files to be bundled directly with apps [Redistributing Visual C++ Files](https://learn.microsoft.com/en-us/cpp/windows/redistributing-visual-cpp-files?view=msvc-170#install-individual-redistributable-files).

However, because of the GPL licensing with JPEGView by the original author, [Is dynamically linking my program with the Visual C++ (or Visual Basic) runtime library permitted under the GPL](https://www.gnu.org/licenses/gpl-faq.html#WindowsRuntimeAndGPL) says
> You may not distribute these libraries in compiled DLL form with the program

So we have to take option 2 or 3.  (I think) we can bundle the redistributable installer with the JPEGView installer.  

With the delayed loading will there be an error when launching JPEGView without the distributable available?  Or does it just do nothing?

sylikc commented:
either way @qbnu thanks for putting in all this work to fix bugs and add support for both the animated webp and jpeg XL.  It's really really cool

qbnu commented:
It shows an error like this when trying to open a JXL file, but you can just click OK and continue using it.
![msvcp-error](https://user-images.githubusercontent.com/93988953/208322512-a88ad19f-0728-4992-8119-760f38ae7dc7.PNG)

Hrxn commented:
Personally, I'd go with that option..
> Tell the user to go download it if they want JPEG XL support

sylikc commented:
Any ideas on how we would have the dll not load (aka no visible error message, or a one time pop-up "you don't have dependencies support for jpeg xl, feature will be disabled") if it's not found?

Usually an error like that for the average user will just freak them out. 🙄

qbnu commented:
We can suppress the pop-up with [SetErrorMode](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-seterrormode), and maybe add an extra error message for JXL files in HelpersGUI.cpp

sylikc commented:
@qbnu should I reopen this or is there more updated code that uses libjxl which we could look at and continue to discuss adding XL support?
